File: databases/meilisearch-patterns/assets/search-client.py
"""
Meilisearch Python Search Client
=================================

A production-ready wrapper around the ``meilisearch`` Python SDK providing:

* Client initialisation with connection validation
* Index & document management (CRUD, bulk import with batching)
* Search helpers (basic, filtered, faceted, geo, hybrid)
* Multi-search across indexes
* Task monitoring & waiting
* Backup utilities (dump creation)
* Tenant-token generation
* CLI interface for common operations

Usage as a library::

    from search_client import MeiliClient

    client = MeiliClient("http://localhost:7700", "masterKey")
    client.add_documents("products", [{"id": 1, "name": "Keyboard"}])
    results = client.search("products", "keyboard")

Usage from the command line::

    python search_client.py search products "mechanical keyboard"
    python search_client.py import products products.json --batch-size 500
    python search_client.py backup
"""

from __future__ import annotations

import argparse
import json
import logging
import sys
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Optional, Sequence

import meilisearch
from meilisearch.errors import MeilisearchApiError, MeilisearchError

# ---------------------------------------------------------------------------
# Logging
# ---------------------------------------------------------------------------

logger = logging.getLogger("meilisearch_client")


def configure_logging(level: int = logging.INFO) -> None:
    """Set up structured logging for the client."""
    handler = logging.StreamHandler(sys.stdout)
    handler.setFormatter(
        logging.Formatter(
            "%(asctime)s [%(levelname)s] %(name)s – %(message)s",
            datefmt="%Y-%m-%d %H:%M:%S",
        )
    )
    logger.addHandler(handler)
    logger.setLevel(level)


# ---------------------------------------------------------------------------
# Custom Exceptions
# ---------------------------------------------------------------------------


class ClientError(Exception):
    """Base exception for client-level errors."""


class ConnectionError(ClientError):  # noqa: A001 – intentional shadow
    """Raised when the client cannot reach Meilisearch."""


class IndexNotFoundError(ClientError):
    """Raised when an index does not exist."""


class TaskFailedError(ClientError):
    """Raised when a Meilisearch task terminates with an error."""

    def __init__(self, task_uid: int, error: dict[str, Any]) -> None:
        self.task_uid = task_uid
        self.error = error
        super().__init__(f"Task {task_uid} failed: {error.get('message', error)}")


# ---------------------------------------------------------------------------
# Task Info Helper
# ---------------------------------------------------------------------------


@dataclass
class TaskInfo:
    """Lightweight representation of a Meilisearch task."""

    uid: int
    status: str
    type: str
    index_uid: Optional[str] = None
    error: Optional[dict[str, Any]] = None
    duration: Optional[str] = None
    enqueued_at: Optional[str] = None
    started_at: Optional[str] = None
    finished_at: Optional[str] = None
    details: dict[str, Any] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# Main Client
# ---------------------------------------------------------------------------


class MeiliClient:
    """High-level Meilisearch client with convenience methods."""

    def __init__(
        self,
        host: str = "http://localhost:7700",
        api_key: Optional[str] = None,
        *,
        timeout: int = 10,
        validate: bool = True,
    ) -> None:
        """
        Initialise the client.

        Args:
            host: Meilisearch server URL.
            api_key: Master or admin API key.
            timeout: HTTP request timeout in seconds.
            validate: If ``True``, perform a health check on init.
        """
        self._host = host.rstrip("/")
        self._client = meilisearch.Client(self._host, api_key, timeout=timeout)
        if validate:
            self._validate_connection()

    # -- connection ---------------------------------------------------------

    def _validate_connection(self) -> None:
        """Verify the server is reachable and healthy."""
        try:
            health = self._client.health()
            if health.get("status") != "available":
                raise ConnectionError(f"Unexpected health status: {health}")
            logger.info("Connected to Meilisearch at %s", self._host)
        except MeilisearchError as exc:
            raise ConnectionError(f"Cannot reach Meilisearch: {exc}") from exc

    @property
    def raw(self) -> meilisearch.Client:
        """Return the underlying SDK client for advanced usage."""
        return self._client

    # -----------------------------------------------------------------------
    # Index Management
    # -----------------------------------------------------------------------

    def create_index(
        self, uid: str, primary_key: Optional[str] = None
    ) -> dict[str, Any]:
        """Create an index and wait for the task to complete."""
        task = self._client.create_index(uid, {"primaryKey": primary_key})
        return self.wait_for_task(task.task_uid)

    def delete_index(self, uid: str) -> dict[str, Any]:
        """Delete an index and wait for the task to complete."""
        task = self._client.delete_index(uid)
        return self.wait_for_task(task.task_uid)

    def list_indexes(self) -> list[dict[str, Any]]:
        """Return metadata for every index on the instance."""
        result = self._client.get_indexes()
        return [
            {"uid": idx.uid, "primaryKey": idx.primary_key, "createdAt": idx.created_at}
            for idx in (result["results"] if isinstance(result, dict) else result)
        ]

    def get_index_stats(self, uid: str) -> dict[str, Any]:
        """Return stats (document count, field distribution, etc.) for an index."""
        index = self._client.index(uid)
        return index.get_stats()

    def update_index_settings(
        self, uid: str, settings: dict[str, Any]
    ) -> dict[str, Any]:
        """Apply partial settings to an index and wait for completion."""
        index = self._client.index(uid)
        task = index.update_settings(settings)
        return self.wait_for_task(task.task_uid)

    def get_index_settings(self, uid: str) -> dict[str, Any]:
        """Get the current settings of an index."""
        index = self._client.index(uid)
        return index.get_settings()

    # -----------------------------------------------------------------------
    # Document Management
    # -----------------------------------------------------------------------

    def add_documents(
        self,
        uid: str,
        documents: list[dict[str, Any]],
        primary_key: Optional[str] = None,
    ) -> dict[str, Any]:
        """Add or replace documents and wait for the task to complete."""
        index = self._client.index(uid)
        task = index.add_documents(documents, primary_key)
        return self.wait_for_task(task.task_uid)

    def update_documents(
        self,
        uid: str,
        documents: list[dict[str, Any]],
        primary_key: Optional[str] = None,
    ) -> dict[str, Any]:
        """Partially update documents (merge fields) and wait."""
        index = self._client.index(uid)
        task = index.update_documents(documents, primary_key)
        return self.wait_for_task(task.task_uid)

    def delete_documents(self, uid: str, ids: list[str | int]) -> dict[str, Any]:
        """Delete documents by primary-key values and wait."""
        index = self._client.index(uid)
        task = index.delete_documents(ids)
        return self.wait_for_task(task.task_uid)

    def get_document(self, uid: str, document_id: str | int) -> dict[str, Any]:
        """Fetch a single document by its primary key."""
        index = self._client.index(uid)
        return index.get_document(document_id)

    def get_documents(
        self,
        uid: str,
        *,
        offset: int = 0,
        limit: int = 20,
        fields: Optional[list[str]] = None,
    ) -> dict[str, Any]:
        """Fetch a paginated list of documents."""
        index = self._client.index(uid)
        params: dict[str, Any] = {"offset": offset, "limit": limit}
        if fields:
            params["fields"] = fields
        return index.get_documents(params)

    def bulk_import(
        self,
        uid: str,
        documents: list[dict[str, Any]],
        *,
        batch_size: int = 1000,
        primary_key: Optional[str] = None,
    ) -> list[dict[str, Any]]:
        """
        Import a large list of documents in batches.

        Returns a list of completed task results – one per batch.
        """
        index = self._client.index(uid)
        results: list[dict[str, Any]] = []
        total = len(documents)

        for start in range(0, total, batch_size):
            batch = documents[start : start + batch_size]
            task = index.add_documents(batch, primary_key)
            result = self.wait_for_task(task.task_uid)
            results.append(result)
            logger.info(
                "Imported batch %d–%d / %d",
                start + 1,
                min(start + batch_size, total),
                total,
            )

        return results

    # -----------------------------------------------------------------------
    # Search
    # -----------------------------------------------------------------------

    def search(
        self,
        uid: str,
        query: str,
        *,
        limit: int = 20,
        offset: int = 0,
        filter: Optional[str | list[str]] = None,  # noqa: A002
        facets: Optional[list[str]] = None,
        sort: Optional[list[str]] = None,
        attributes_to_retrieve: Optional[list[str]] = None,
        attributes_to_highlight: Optional[list[str]] = None,
        page: Optional[int] = None,
        hits_per_page: Optional[int] = None,
    ) -> dict[str, Any]:
        """
        Execute a search query with optional filters, facets, and sorting.

        Args:
            uid: Index identifier.
            query: Full-text query string.
            limit: Maximum number of hits (offset-based pagination).
            offset: Starting offset (offset-based pagination).
            filter: Filter expression string or array.
            facets: Attributes to compute facet counts for.
            sort: Sort rules, e.g. ``["price:asc"]``.
            attributes_to_retrieve: Fields to include in each hit.
            attributes_to_highlight: Fields to highlight with match tags.
            page: Page number for page-based pagination (1-indexed).
            hits_per_page: Hits per page for page-based pagination.

        Returns:
            Raw search response dict (hits, query, processingTimeMs, …).
        """
        index = self._client.index(uid)
        params: dict[str, Any] = {}

        if page is not None:
            params["page"] = page
            if hits_per_page is not None:
                params["hitsPerPage"] = hits_per_page
        else:
            params["limit"] = limit
            params["offset"] = offset

        if filter:
            params["filter"] = filter
        if facets:
            params["facets"] = facets
        if sort:
            params["sort"] = sort
        if attributes_to_retrieve:
            params["attributesToRetrieve"] = attributes_to_retrieve
        if attributes_to_highlight:
            params["attributesToHighlight"] = attributes_to_highlight

        return index.search(query, params)

    def geo_search(
        self,
        uid: str,
        query: str,
        *,
        lat: float,
        lng: float,
        radius_m: Optional[int] = None,
        sort_by_distance: bool = True,
        limit: int = 20,
    ) -> dict[str, Any]:
        """
        Search with geo-radius filtering and optional distance sorting.

        Args:
            uid: Index identifier.
            query: Full-text query.
            lat: Latitude of the reference point.
            lng: Longitude of the reference point.
            radius_m: Maximum distance in metres. ``None`` disables the filter.
            sort_by_distance: Sort results by distance ascending.
            limit: Maximum number of hits.
        """
        params: dict[str, Any] = {"limit": limit}

        if radius_m is not None:
            params["filter"] = f"_geoRadius({lat}, {lng}, {radius_m})"
        if sort_by_distance:
            params["sort"] = [f"_geoPoint({lat}, {lng}):asc"]

        index = self._client.index(uid)
        return index.search(query, params)

    def hybrid_search(
        self,
        uid: str,
        query: str,
        *,
        semantic_ratio: float = 0.5,
        embedder: str = "default",
        limit: int = 20,
    ) -> dict[str, Any]:
        """
        Perform hybrid keyword + semantic search (Meilisearch v1.3+).

        Args:
            uid: Index identifier.
            query: Natural-language query.
            semantic_ratio: 0.0 = pure keyword, 1.0 = pure semantic.
            embedder: Name of the configured embedder.
            limit: Maximum hits.
        """
        index = self._client.index(uid)
        params: dict[str, Any] = {
            "limit": limit,
            "hybrid": {
                "semanticRatio": semantic_ratio,
                "embedder": embedder,
            },
        }
        return index.search(query, params)

    def multi_search(self, queries: list[dict[str, Any]]) -> dict[str, Any]:
        """
        Execute multiple search queries in a single HTTP request.

        Args:
            queries: List of dicts, each with ``indexUid``, ``q``, and
                     optional search parameters.

        Returns:
            Combined results keyed by index.
        """
        return self._client.multi_search(queries)

    # -----------------------------------------------------------------------
    # Task Monitoring
    # -----------------------------------------------------------------------

    def wait_for_task(
        self,
        task_uid: int,
        *,
        timeout_ms: int = 30_000,
        interval_ms: int = 250,
    ) -> dict[str, Any]:
        """
        Poll until a task reaches a terminal state.

        Raises:
            TaskFailedError: If the task fails.
            TimeoutError: If the timeout is exceeded.
        """
        deadline = time.monotonic() + timeout_ms / 1000
        while True:
            task = self._client.get_task(task_uid)
            status = task.get("status") if isinstance(task, dict) else getattr(task, "status", None)

            if status in ("succeeded", "failed", "canceled"):
                task_dict = task if isinstance(task, dict) else vars(task)
                if status == "failed":
                    error = task_dict.get("error", {})
                    raise TaskFailedError(task_uid, error)
                return task_dict

            if time.monotonic() > deadline:
                raise TimeoutError(
                    f"Task {task_uid} did not complete within {timeout_ms}ms"
                )
            time.sleep(interval_ms / 1000)

    def get_task(self, task_uid: int) -> TaskInfo:
        """Retrieve details for a single task."""
        raw = self._client.get_task(task_uid)
        t = raw if isinstance(raw, dict) else vars(raw)
        return TaskInfo(
            uid=t["uid"],
            status=t["status"],
            type=t["type"],
            index_uid=t.get("indexUid"),
            error=t.get("error"),
            duration=t.get("duration"),
            enqueued_at=t.get("enqueuedAt"),
            started_at=t.get("startedAt"),
            finished_at=t.get("finishedAt"),
            details=t.get("details", {}),
        )

    def list_tasks(
        self,
        *,
        index_uids: Optional[list[str]] = None,
        statuses: Optional[list[str]] = None,
        types: Optional[list[str]] = None,
        limit: int = 20,
    ) -> list[TaskInfo]:
        """List recent tasks with optional filters."""
        params: dict[str, Any] = {"limit": limit}
        if index_uids:
            params["indexUids"] = index_uids
        if statuses:
            params["statuses"] = statuses
        if types:
            params["types"] = types

        result = self._client.get_tasks(params)
        raw_list = result.get("results", []) if isinstance(result, dict) else result
        return [
            TaskInfo(
                uid=t.get("uid", t.get("taskUid", 0)),
                status=t.get("status", "unknown"),
                type=t.get("type", "unknown"),
                index_uid=t.get("indexUid"),
                error=t.get("error"),
                duration=t.get("duration"),
            )
            for t in raw_list
        ]

    # -----------------------------------------------------------------------
    # Backup / Dumps
    # -----------------------------------------------------------------------

    def create_dump(self) -> dict[str, Any]:
        """Trigger a full database dump and wait for it to finish."""
        task = self._client.create_dump()
        uid = task.task_uid if hasattr(task, "task_uid") else task.get("taskUid", task.get("uid"))
        return self.wait_for_task(uid, timeout_ms=300_000)

    # -----------------------------------------------------------------------
    # Tenant Tokens
    # -----------------------------------------------------------------------

    def generate_tenant_token(
        self,
        api_key_uid: str,
        search_rules: dict[str, Any] | list[str],
        *,
        expires_at: Optional[str] = None,
        api_key: Optional[str] = None,
    ) -> str:
        """
        Generate a tenant token (JWT) embedding search-rule restrictions.

        Args:
            api_key_uid: UID of the parent API key (from ``/keys``).
            search_rules: Per-index filter rules or list of allowed indexes.
            expires_at: ISO-8601 expiry timestamp.
            api_key: Signing key. Defaults to the client's key.

        Returns:
            Signed JWT string.
        """
        return self._client.generate_tenant_token(
            api_key_uid=api_key_uid,
            search_rules=search_rules,
            expires_at=expires_at,
            api_key=api_key,
        )


# ---------------------------------------------------------------------------
# CLI Interface
# ---------------------------------------------------------------------------


def _build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        prog="search-client",
        description="Meilisearch CLI – common operations from the command line.",
    )
    parser.add_argument(
        "--host",
        default="http://localhost:7700",
        help="Meilisearch URL (default: %(default)s)",
    )
    parser.add_argument("--api-key", default=None, help="API key")
    parser.add_argument(
        "-v", "--verbose", action="store_true", help="Enable debug logging"
    )

    sub = parser.add_subparsers(dest="command", required=True)

    # -- search -------------------------------------------------------------
    sp_search = sub.add_parser("search", help="Run a search query")
    sp_search.add_argument("index", help="Index UID")
    sp_search.add_argument("query", help="Search query")
    sp_search.add_argument("--limit", type=int, default=10)
    sp_search.add_argument("--filter", default=None)
    sp_search.add_argument("--sort", nargs="*", default=None)
    sp_search.add_argument("--facets", nargs="*", default=None)

    # -- import -------------------------------------------------------------
    sp_import = sub.add_parser("import", help="Import documents from a JSON file")
    sp_import.add_argument("index", help="Index UID")
    sp_import.add_argument("file", help="Path to JSON file (array of objects)")
    sp_import.add_argument("--batch-size", type=int, default=1000)
    sp_import.add_argument("--primary-key", default=None)

    # -- backup -------------------------------------------------------------
    sub.add_parser("backup", help="Create a full database dump")

    # -- indexes ------------------------------------------------------------
    sub.add_parser("indexes", help="List all indexes")

    # -- settings -----------------------------------------------------------
    sp_settings = sub.add_parser(
        "settings", help="Get or apply index settings"
    )
    sp_settings.add_argument("index", help="Index UID")
    sp_settings.add_argument(
        "--apply",
        default=None,
        help="Path to a JSON settings file to apply",
    )

    # -- tasks --------------------------------------------------------------
    sp_tasks = sub.add_parser("tasks", help="List recent tasks")
    sp_tasks.add_argument("--limit", type=int, default=10)

    return parser


def _run_cli(argv: Sequence[str] | None = None) -> None:
    parser = _build_parser()
    args = parser.parse_args(argv)

    configure_logging(logging.DEBUG if args.verbose else logging.INFO)

    client = MeiliClient(args.host, args.api_key, validate=True)

    if args.command == "search":
        results = client.search(
            args.index,
            args.query,
            limit=args.limit,
            filter=args.filter,
            sort=args.sort,
            facets=args.facets,
        )
        print(json.dumps(results, indent=2, default=str))

    elif args.command == "import":
        data = json.loads(Path(args.file).read_text())
        if not isinstance(data, list):
            logger.error("JSON file must contain an array of objects.")
            sys.exit(1)
        results = client.bulk_import(
            args.index, data, batch_size=args.batch_size, primary_key=args.primary_key
        )
        print(f"Imported {len(data)} documents in {len(results)} batch(es).")

    elif args.command == "backup":
        result = client.create_dump()
        print(f"Dump completed. Task: {json.dumps(result, indent=2, default=str)}")

    elif args.command == "indexes":
        indexes = client.list_indexes()
        print(json.dumps(indexes, indent=2, default=str))

    elif args.command == "settings":
        if args.apply:
            settings = json.loads(Path(args.apply).read_text())
            result = client.update_index_settings(args.index, settings)
            print(f"Settings applied. Task: {json.dumps(result, indent=2, default=str)}")
        else:
            settings = client.get_index_settings(args.index)
            print(json.dumps(settings, indent=2, default=str))

    elif args.command == "tasks":
        tasks = client.list_tasks(limit=args.limit)
        for t in tasks:
            print(f"[{t.status:>10}] #{t.uid}  {t.type}  index={t.index_uid}")


# ---------------------------------------------------------------------------
# Entry Point
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    _run_cli()

File: databases/meilisearch-patterns/assets/search-client.ts
/**
 * Meilisearch TypeScript Search Client
 *
 * A type-safe wrapper around the Meilisearch JS SDK providing:
 *  - Generic index operations
 *  - Filtered, faceted, sorted, and paginated search
 *  - InstantSearch.js adapter integration
 *  - Hybrid / semantic search helpers
 *  - Multi-index search
 *  - Tenant-token generation
 *  - Document CRUD & task waiting utilities
 *  - Index settings management
 *
 * @example
 * ```ts
 * import { createClient, search, addDocuments } from "./search-client";
 *
 * const client = createClient("http://localhost:7700", "masterKey");
 * await addDocuments(client, "products", [{ id: 1, name: "Keyboard" }]);
 * const results = await search<Product>(client, "products", "keyboard");
 * ```
 */

import {
  MeiliSearch,
  Index,
  SearchParams,
  SearchResponse,
  MultiSearchParams,
  MultiSearchResponse,
  EnqueuedTask,
  Task,
  Settings,
  TokenSearchRules,
} from "meilisearch";

// Re-export the SDK client type for consumers.
export type { MeiliSearch, Index, SearchResponse, Task, Settings };

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Options accepted by {@link createClient}. */
export interface ClientOptions {
  /** Meilisearch host URL (default: http://localhost:7700). */
  host?: string;
  /** API key – use a search-only key in the browser. */
  apiKey?: string;
  /** Custom request headers forwarded with every HTTP call. */
  headers?: Record<string, string>;
}

/** Extended search parameters exposing the most useful knobs. */
export interface SearchOptions<T = Record<string, unknown>> {
  /** Full-text query string. */
  q?: string;
  /** Filter expression (string or nested array). */
  filter?: string | string[] | string[][];
  /** Attributes to use as facets in the response. */
  facets?: string[];
  /** Sort rules, e.g. ["price:asc", "rating:desc"]. */
  sort?: string[];
  /** Number of results per page (default 20). */
  limit?: number;
  /** Offset for cursor-based pagination. */
  offset?: number;
  /** Page number (1-based) for page-based pagination. */
  page?: number;
  /** Hits per page when using page-based pagination. */
  hitsPerPage?: number;
  /** Attributes to retrieve (defaults to displayedAttributes). */
  attributesToRetrieve?: (keyof T & string)[];
  /** Attributes to highlight. */
  attributesToHighlight?: (keyof T & string)[];
  /** Attributes to crop. */
  attributesToCrop?: (keyof T & string)[];
  /** Enable hybrid search with semantic ratio (0 = keyword, 1 = semantic). */
  hybridSemanticRatio?: number;
  /** Embedder to use for hybrid search. */
  hybridEmbedder?: string;
}

/** Typed error wrapper for Meilisearch API errors. */
export class MeiliSearchError extends Error {
  constructor(
    message: string,
    public readonly code: string,
    public readonly httpStatus: number,
    public readonly link?: string,
  ) {
    super(message);
    this.name = "MeiliSearchError";
  }
}

// ---------------------------------------------------------------------------
// Client Initialization
// ---------------------------------------------------------------------------

/**
 * Create and return a configured MeiliSearch client instance.
 *
 * @param host - Meilisearch server URL.
 * @param apiKey - API key for authentication.
 * @param headers - Optional custom HTTP headers.
 * @returns A {@link MeiliSearch} client.
 */
export function createClient(
  host = "http://localhost:7700",
  apiKey?: string,
  headers?: Record<string, string>,
): MeiliSearch {
  return new MeiliSearch({ host, apiKey, headers });
}

// ---------------------------------------------------------------------------
// Index Operations
// ---------------------------------------------------------------------------

/**
 * Get or create an index, returning a typed {@link Index} handle.
 *
 * @param client - MeiliSearch client instance.
 * @param uid - Unique index identifier.
 * @param primaryKey - Optional primary key field name.
 * @returns The index handle.
 */
export async function getOrCreateIndex<T extends Record<string, unknown>>(
  client: MeiliSearch,
  uid: string,
  primaryKey?: string,
): Promise<Index<T>> {
  try {
    return await client.getIndex<T>(uid);
  } catch {
    const task = await client.createIndex(uid, { primaryKey });
    await client.waitForTask(task.taskUid);
    return client.getIndex<T>(uid);
  }
}

/**
 * Delete an index by its uid.
 *
 * @param client - MeiliSearch client instance.
 * @param uid - Index identifier to delete.
 */
export async function deleteIndex(
  client: MeiliSearch,
  uid: string,
): Promise<void> {
  const task = await client.deleteIndex(uid);
  await client.waitForTask(task.taskUid);
}

// ---------------------------------------------------------------------------
// Search
// ---------------------------------------------------------------------------

/**
 * Perform a search on the given index with type-safe options.
 *
 * @typeParam T - Shape of a document in the index.
 * @param client - MeiliSearch client.
 * @param indexUid - Target index identifier.
 * @param query - Search query string.
 * @param options - Additional search options.
 * @returns Typed search response.
 */
export async function search<T extends Record<string, unknown>>(
  client: MeiliSearch,
  indexUid: string,
  query: string,
  options: SearchOptions<T> = {},
): Promise<SearchResponse<T>> {
  const index = client.index<T>(indexUid);

  const params: SearchParams = {
    filter: options.filter,
    facets: options.facets,
    sort: options.sort,
    limit: options.limit,
    offset: options.offset,
    page: options.page,
    hitsPerPage: options.hitsPerPage,
    attributesToRetrieve: options.attributesToRetrieve as string[],
    attributesToHighlight: options.attributesToHighlight as string[],
    attributesToCrop: options.attributesToCrop as string[],
  };

  // Hybrid / semantic search support (Meilisearch v1.3+).
  if (options.hybridSemanticRatio !== undefined) {
    (params as Record<string, unknown>).hybrid = {
      semanticRatio: options.hybridSemanticRatio,
      embedder: options.hybridEmbedder ?? "default",
    };
  }

  return index.search(query, params);
}

/**
 * Execute a multi-index search in a single HTTP round-trip.
 *
 * @param client - MeiliSearch client.
 * @param queries - Array of per-index search queries.
 * @returns Combined results from all queried indexes.
 */
export async function multiSearch(
  client: MeiliSearch,
  queries: MultiSearchParams["queries"],
): Promise<MultiSearchResponse> {
  return client.multiSearch({ queries });
}

/**
 * Helper for hybrid / semantic search on a single index.
 *
 * @param client - MeiliSearch client.
 * @param indexUid - Target index.
 * @param query - Natural-language query.
 * @param semanticRatio - Balance between keyword (0) and semantic (1). Default 0.5.
 * @param embedder - Name of the configured embedder. Default "default".
 * @returns Search response.
 */
export async function hybridSearch<T extends Record<string, unknown>>(
  client: MeiliSearch,
  indexUid: string,
  query: string,
  semanticRatio = 0.5,
  embedder = "default",
): Promise<SearchResponse<T>> {
  return search<T>(client, indexUid, query, {
    hybridSemanticRatio: semanticRatio,
    hybridEmbedder: embedder,
  });
}

// ---------------------------------------------------------------------------
// InstantSearch.js Integration
// ---------------------------------------------------------------------------

/**
 * Create an InstantSearch.js-compatible search client using the
 * `@meilisearch/instant-meilisearch` adapter.
 *
 * @example
 * ```ts
 * import { createInstantSearchClient } from "./search-client";
 * import instantsearch from "instantsearch.js";
 * import { searchBox, hits } from "instantsearch.js/es/widgets";
 *
 * const { searchClient, setMeiliSearchParams } = createInstantSearchClient(
 *   "http://localhost:7700",
 *   "searchOnlyApiKey",
 * );
 *
 * const search = instantsearch({ indexName: "products", searchClient });
 * search.addWidgets([searchBox({ container: "#search" }), hits({ container: "#hits" })]);
 * search.start();
 * ```
 *
 * @param host - Meilisearch URL.
 * @param apiKey - Search-only API key.
 * @param options - Adapter-specific options (finitePagination, etc.).
 * @returns The adapter's return value (searchClient + helpers).
 */
export function createInstantSearchClient(
  host: string,
  apiKey: string,
  options: Record<string, unknown> = {},
) {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { instantMeiliSearch } = require("@meilisearch/instant-meilisearch");
  return instantMeiliSearch(host, apiKey, {
    finitePagination: true,
    ...options,
  });
}

// ---------------------------------------------------------------------------
// Document CRUD
// ---------------------------------------------------------------------------

/**
 * Add or replace documents in an index.
 *
 * @param client - MeiliSearch client.
 * @param indexUid - Target index.
 * @param documents - Array of document objects.
 * @param primaryKey - Optional primary key override.
 * @returns The enqueued task.
 */
export async function addDocuments<T extends Record<string, unknown>>(
  client: MeiliSearch,
  indexUid: string,
  documents: T[],
  primaryKey?: string,
): Promise<EnqueuedTask> {
  const index = client.index<T>(indexUid);
  return index.addDocuments(documents, { primaryKey });
}

/**
 * Partially update documents (merge with existing fields).
 *
 * @param client - MeiliSearch client.
 * @param indexUid - Target index.
 * @param documents - Partial document objects (must contain the primary key).
 * @returns The enqueued task.
 */
export async function updateDocuments<T extends Record<string, unknown>>(
  client: MeiliSearch,
  indexUid: string,
  documents: Partial<T>[],
): Promise<EnqueuedTask> {
  const index = client.index<T>(indexUid);
  return index.updateDocuments(documents as T[]);
}

/**
 * Delete documents by their primary-key values.
 *
 * @param client - MeiliSearch client.
 * @param indexUid - Target index.
 * @param ids - Array of document IDs (primary key values).
 * @returns The enqueued task.
 */
export async function deleteDocuments(
  client: MeiliSearch,
  indexUid: string,
  ids: (string | number)[],
): Promise<EnqueuedTask> {
  const index = client.index(indexUid);
  return index.deleteDocuments(ids);
}

// ---------------------------------------------------------------------------
// Task Utilities
// ---------------------------------------------------------------------------

/**
 * Wait for a task to reach a terminal state (succeeded / failed).
 *
 * @param client - MeiliSearch client.
 * @param taskUid - Task identifier returned by any write operation.
 * @param timeoutMs - Maximum time to wait in milliseconds (default 30 000).
 * @param intervalMs - Polling interval in milliseconds (default 250).
 * @returns The completed {@link Task}.
 * @throws {MeiliSearchError} If the task fails.
 */
export async function waitForTask(
  client: MeiliSearch,
  taskUid: number,
  timeoutMs = 30_000,
  intervalMs = 250,
): Promise<Task> {
  const task = await client.waitForTask(taskUid, {
    timeOutMs: timeoutMs,
    intervalMs,
  });

  if (task.status === "failed") {
    throw new MeiliSearchError(
      task.error?.message ?? "Task failed",
      task.error?.code ?? "unknown",
      0,
      task.error?.link,
    );
  }

  return task;
}

// ---------------------------------------------------------------------------
// Index Settings Management
// ---------------------------------------------------------------------------

/**
 * Retrieve the full settings object for an index.
 *
 * @param client - MeiliSearch client.
 * @param indexUid - Target index.
 * @returns Current index {@link Settings}.
 */
export async function getSettings(
  client: MeiliSearch,
  indexUid: string,
): Promise<Settings> {
  const index = client.index(indexUid);
  return index.getSettings();
}

/**
 * Apply a partial settings update to an index and wait for completion.
 *
 * @param client - MeiliSearch client.
 * @param indexUid - Target index.
 * @param settings - Partial {@link Settings} to merge.
 * @returns The completed task.
 */
export async function updateSettings(
  client: MeiliSearch,
  indexUid: string,
  settings: Settings,
): Promise<Task> {
  const index = client.index(indexUid);
  const enqueued = await index.updateSettings(settings);
  return waitForTask(client, enqueued.taskUid);
}

/**
 * Reset all settings of an index to their default values.
 *
 * @param client - MeiliSearch client.
 * @param indexUid - Target index.
 * @returns The completed task.
 */
export async function resetSettings(
  client: MeiliSearch,
  indexUid: string,
): Promise<Task> {
  const index = client.index(indexUid);
  const enqueued = await index.resetSettings();
  return waitForTask(client, enqueued.taskUid);
}

// ---------------------------------------------------------------------------
// Tenant Token Generation
// ---------------------------------------------------------------------------

/**
 * Generate a tenant token that restricts search to specific filter rules.
 *
 * Tenant tokens are JWTs signed with an API key; they let you expose
 * search to end-users while enforcing row-level security.
 *
 * @param client - MeiliSearch client.
 * @param apiKeyUid - UID of the parent API key (visible in /keys).
 * @param searchRules - Per-index filter rules to embed in the token.
 * @param expiresAt - Optional expiry date for the token.
 * @returns A signed JWT string.
 *
 * @example
 * ```ts
 * const token = await generateTenantToken(client, apiKeyUid, {
 *   products: { filter: "tenant_id = 42" },
 * });
 * ```
 */
export async function generateTenantToken(
  client: MeiliSearch,
  apiKeyUid: string,
  searchRules: TokenSearchRules,
  expiresAt?: Date,
): Promise<string> {
  return client.generateTenantToken(apiKeyUid, searchRules, {
    expiresAt,
  });
}

// ---------------------------------------------------------------------------
// Error Handling Utility
// ---------------------------------------------------------------------------

/**
 * Wrap an async Meilisearch call, converting SDK errors into {@link MeiliSearchError}.
 *
 * @param fn - Async function to execute.
 * @returns The resolved value of `fn`.
 * @throws {MeiliSearchError} on failure.
 */
export async function withErrorHandling<T>(fn: () => Promise<T>): Promise<T> {
  try {
    return await fn();
  } catch (err: unknown) {
    if (err && typeof err === "object" && "httpStatus" in err) {
      const e = err as {
        message: string;
        code: string;
        httpStatus: number;
        link?: string;
      };
      throw new MeiliSearchError(e.message, e.code, e.httpStatus, e.link);
    }
    throw err;
  }
}
